### ISSUE TITLE
[question] Matching on maps not supported?

### ISSUE THREAD

**Issue by nietaki** [massemanet/redbug#2]
When playing around with `redbug` (which I loved right out of the gate) I realized it doesn't support matching maps in the trace pattern arguments:
```Erlang
10> redbug:start("erlang:term_to_binary([])").
{38,1}
redbug done, timeout - 0
11> [A, B, C] = "#\{\}".
"#{}"
12> redbug:start("erlang:term_to_binary(#\{\})").
{bad_type,{map,[]}}
13> redbug:start("erlang:term_to_binary(X) when X == 1").
{38,1}
14> redbug:stop().
stopped
redbug done, local_done - 0
15> redbug:start("erlang:term_to_binary(X) when #\{\} = X").
{function_clause,{parse_guards,"erlang:term_to_binary(X) when #{} = X"},
                 [{redbug_msc,arg,
                              [{match,1,{map,1,[]},{var,1,'X'}}],
                              (... more errors cut)
```
After looking at the source code, I think it's intentional.

While I'm not questioning the decision, I'm curious to the motivation. Is there any reason in particular why matching maps isn't supported in the trace pattern?

Thanks in advance and sorry to bother you with this GitHub issue.

**Comment by massemanet** [comment]
sorry for the late reply.

AFAICT, the BEAM tracing does not support matching on maps.

Possibly that has changed in erlang 20?

**Comment by nietaki** [comment]
I couldn't find any information on it, so today I tried to repro it to see if it's supported. It seems like it's possible, at least on Erlang 20.

`wat.erl`:
```erlang
-module(wat).
-export([start/1]).

start(X) -> X.
```

```erlang
Erlang/OTP 20 [erts-9.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V9.0  (abort with ^G)
1> c(wat).
{ok,wat}
2> l(wat).
{module,wat}
3> erlang:trace(all, true, [call]).
29
4> erlang:trace_pattern({wat, start, '_'}, [{[#{foo => bar}], [], []}], [local]).
1
5> spawn(fun() -> wat:start(#{baz => ban}) end).
<0.70.0>
6> spawn(fun() -> wat:start(#{foo => bar}) end).
<0.72.0>
7> spawn(fun() -> wat:start(foo) end).
<0.74.0>
8> flush().
Shell got {trace,<0.72.0>,call,{wat,start,[#{foo => bar}]}}
ok
```

I'll try earlier versions of Erlang to see where it's possible.

I'd be happy to write a PR to make matching on maps possible in `redbug`, but considering my understanding on the low level tracing is shaky at best and I don't write Erlang day to day, you might be able to do it much quicker and with better code quality 😄

**Comment by nietaki** [comment]
The example I posted works in in 19.3 and 18.3, I had problems compiling 17.x so I couldn't check further than that.

**Comment by massemanet** [comment]
I'll do it. I've been waiting for them to support maps (or maybe they did all along and I just missed it).
thanks for the help!

**Comment by nietaki** [comment]
Brilliant, thank you!

**Comment by massemanet** [comment]
(Possibly) fixed in 0bc3ffa

**Comment by nietaki** [comment]
I tried 935ae87e557621033d1a0a51fc03b8bfe15a2bee and it works like a charm, with var binding, good errors for invalid trace patterns and everything.

From CI it looks like it doesn't work with Erlang 16 anymore, but it should be ok with semver if you release it under 1.2.x.

Thank you for doing this!

**Comment by massemanet** [comment]
I'll tag it with 1.2.0
thx for testing.

**Comment by massemanet** [comment]
fixed in 1.2.0
